x-pack/legacy/plugins/siem/server/lib/detection_engine/routes/rules/import_rules_route.test.ts

```diff
@@ -0,0 +1,399 @@
+        const parsedRequest2: ImportSuccessError = JSON.parse(payloadRequest2);
+
+        expect(parsedRequest2).toEqual({
+          errors: [
+            {
+              error: {
+                message: 'rule_id: "rule-1" already exists',
+                status_code: 409,
+              },
+              rule_id: 'rule-1',
+            },
+          ],
+          success: false,
+          success_count: 0,
+        });
+        expect(statusCodeRequest2).toEqual(200);
+      });
+
+      test('returns with NO reported conflict if `overwrite` is set to `true`', async () => {
+        clients.alertsClient.find.mockResolvedValue(getFindResult());
+
+        const requestPayload = getSimpleRuleAsMultipartContent(['rule-1']);
+        const { statusCode, payload } = await server.inject(getImportRulesRequest(requestPayload));
+        const parsed: ImportSuccessError = JSON.parse(payload);
+
+        expect(parsed).toEqual({
+          errors: [],
+          success: true,
+          success_count: 1,
+        });
+        expect(statusCode).toEqual(200);
+
+        clients.alertsClient.find.mockResolvedValue(getFindResultWithSingleHit());
+        clients.alertsClient.get.mockResolvedValue(getResult());
+
+        const { statusCode: statusCodeRequest2, payload: payloadRequest2 } = await server.inject(
+          getImportRulesRequestOverwriteTrue(requestPayload)
+        );
+        const parsedRequest2: ImportSuccessError = JSON.parse(payloadRequest2);
+
+        expect(parsedRequest2).toEqual({
+          errors: [],
+          success: true,
+          success_count: 1,
+        });
+        expect(statusCodeRequest2).toEqual(200);
+      });
+    });
+  });
+
+  describe('multi rule import', () => {
+    test('returns 200 if all rules imported successfully', async () => {
+      clients.alertsClient.find.mockResolvedValue(getFindResult());
+
+      const requestPayload = getSimpleRuleAsMultipartContent(['rule-1', 'rule-2']);
+      const { statusCode, payload } = await server.inject(getImportRulesRequest(requestPayload));
+      const parsed: ImportSuccessError = JSON.parse(payload);
+
+      expect(parsed).toEqual({
+        errors: [],
+        success: true,
+        success_count: 2,
+      });
+      expect(statusCode).toEqual(200);
+    });
+
+    test('returns 200 with reported conflict if error parsing rule', async () => {
+      const multipartPayload =
+        '--frank_is_awesome\r\n' +
+        `Content-Disposition: form-data; name="file"; filename="rules.ndjson"\r\n` +
+        'Content-Type: application/octet-stream\r\n' +
+        '\r\n' +
+        '{"name"::"Simple Rule Query","description":"Simple Rule Query","risk_score":1,"rule_id":"rule-1","severity":"high","type":"query","query":"user.name: root or user.name: admin"}\r\n' +
+        '{"name":"Simple Rule Query","description":"Simple Rule Query","risk_score":1,"rule_id":"rule-2","severity":"high","type":"query","query":"user.name: root or user.name: admin"}\r\n' +
```

Comment:
Optional:

What I typically do is try to keep as many things `typed` as I can and outside of strings where possible so that if the types change the type checker will catch it first before the tests.

For example I can change this to be this section of code:

```ts
      // This is not parseable due to the extra colon in it: "name"::"
      const unparseableLine =
        '{"name"::"Simple Rule Query","description":"Simple Rule Query","risk_score":1,"rule_id":"rule-1","severity":"high","type":"query","query":"user.name: root or user.name: admin"}';

      const validRule: Partial<OutputRuleAlertRest> = JSON.stringify({
        name: 'Simple Rule Query',
        description: 'Simple Rule Query',
        risk_score: 1,
        rule_id: 'rule-2',
        severity: 'high',
        type: 'query',
        query: 'user.name: root or user.name: admin',
      });

      const multipartPayload =
        '--frank_is_awesome\r\n' +
        `Content-Disposition: form-data; name="file"; filename="rules.ndjson"\r\n` +
        'Content-Type: application/octet-stream\r\n' +
        '\r\n' +
        `${unparseableLine}\r\n` +
        `${validRule}\r\n` +
        '--frank_is_awesome--\r\n';
```

and it makes it easier if I have to change the `validRule` at a later time to be correct again if the format changes as it will be a typed object rather than a string within the test.

Optional though, still really enjoying reading all the tests! They are really nice.

Reply:
I moved a few of these things out into the utils I created after going through your feedback.